### PR TITLE
feat(cockroachdb metrics and logs plugin)

### PR DESCRIPTION
**Pull request by BominRahmani**
The following is a plugin for CockroachDB that collects metrics and logs. This version is strictly for the dedicated/local version of cockroachDB.

# Example config for Metrics
image

# Example config for Logs
image

Example Log files are attached.
[cockroach.log](https://github.com/observIQ/observiq-otel-collector/files/11453007/cockroach.log)
[cockroach-telemetry.log](https://github.com/observIQ/observiq-otel-collector/files/11453013/cockroach-telemetry.log)
[cockroach-pebble.log](https://github.com/observIQ/observiq-otel-collector/files/11453022/cockroach-pebble.log)
[cockroach-kv-distribution.log](https://github.com/observIQ/observiq-otel-collector/files/11453028/cockroach-kv-distribution.log)
[cockroach-health.log](https://github.com/observIQ/observiq-otel-collector/files/11453029/cockroach-health.log)
[cockroach-stderr.log](https://github.com/observIQ/observiq-otel-collector/files/11453041/cockroach-stderr.log)

### CODE REVIEW

File: plugins/cockroachdb_logs.yaml [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+    description: The absolute path to the CockroachDB health logs
+    type: "[]string"
+    default:
+      - "/var/log/cockroach-data/logs/cockroach-health.log"
+  - name: enable_dev_log
+    description: Enable to collect general developer logs.
+    type: bool
+    default: true
+  - name: dev_log_path
+    description: The absolute path to the CockroachDB Dev Logs
```
Comment (dehaansa) [inline plugins/cockroachdb_logs.yaml:19]: Should refer to these as dev logs or developer logs consistently between the different parameters.

File: plugins/cockroachdb_logs.yaml [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+          severity:
+            parse_from: attributes.Type
+    {{ end }} # End of pebble logs.
+
+  service:
+    pipelines:
+      logs:
+        receivers:
+          {{ if .enable_health_log }}
+          - filelog/health_logs
```
Comment (dehaansa) [inline plugins/cockroachdb_logs.yaml:241]: Should all of these named be prefixed with cockroachdb? IE `filelog/cockroachdb_health_logs`

**Review by dehaansa — COMMENTED** [review]
Few comments, overall looks good.

**Review by dehaansa — APPROVED** [review]

**Comment by cpheps** [comment]
@BominRahmani Can you do the following before we review:
- Run `make create-plugin-docs` that should resolve the `check-plugin-docs` step
- Rebase on `release/v1.25.0` branch
- Include samples of the collected logs in the PR description

File: plugins/cockroachdb_logs.yaml [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+    description: At startup, where to start reading logs from the file ('beginning' or 'end')
+    type: string
+    supported:
+      - beginning
+      - end
+    default: beginning
+  - name: timezone
+    description: Timezone to use when parsing the timestamp
+    type: timezone
+    default: UTC
```
Comment (cpheps) [inline plugins/cockroachdb_logs.yaml:78]: Can we add the [offset_storage_dir](https://github.com/observIQ/observiq-otel-collector/blob/release/v1.25.0/plugins/macos_logs.yaml#L30-L33) into here as well?

File: plugins/cockroachdb_logs.yaml [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+      operators:
+        - type: regex_parser
+          regex: '^(?P<Type>[IWEF])(?P<timestamp>\d{6}\s+\d{2}:\d{2}:\d{2}\.\d{6})\s+(?P<GoID>\d+)\s*(?:(?P<Channel>\d+)@)?(?P<Location>\S+)\s+(?:)⋮?\s+(?P<Node>\[[^\]]+\])\s+(?:(?P<Counter>\d+\s+)?)(?P<Message>.*)'
+          timestamp:
+            parse_from: attributes.timestamp
+            layout: '%y%m%d %H:%M:%S.%f'
+            location: {{ .timezone }}
+          severity: 
+            parse_from: attributes.Type
+    {{ end }} ## End Health Logs.
```
Comment (cpheps) [inline plugins/cockroachdb_logs.yaml:101]: nit: remove the `## End health logs`

File: plugins/cockroachdb_logs.yaml [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+      operators:
+        - type: regex_parser
+          regex: '^(?P<Type>[IWEF])(?P<timestamp>\d{6}\s+\d{2}:\d{2}:\d{2}\.\d{6})\s+(?P<GoID>\d+)\s*(?:(?P<Channel>\d+)@)?(?P<Location>\S+)\s+(?:)⋮?\s+(?P<Node>\[[^\]]+\])\s+(?:(?P<Counter>\d+\s+)?)(?P<Message>.*)'
+          timestamp:
+            parse_from: attributes.timestamp
+            layout: '%y%m%d %H:%M:%S.%f'
+            location: {{ .timezone }}
+          severity:
+            parse_from: attributes.Type
+    {{ end }} ## End of dev logs.
```
Comment (cpheps) [inline plugins/cockroachdb_logs.yaml:123]: nit: Remove the comment

File: plugins/cockroachdb_logs.yaml [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+      operators:
+        - type: regex_parser
+          regex: '^(?P<Type>[IWEF])(?P<timestamp>\d{6}\s+\d{2}:\d{2}:\d{2}\.\d{6})\s+(?P<GoID>\d+)\s*(?:(?P<Channel>\d+)@)?(?P<Location>\S+)\s+(?:)⋮?\s+(?P<Node>\[[^\]]+\])\s+(?:(?P<Counter>\d+\s+)?)(?P<Message>.*)'
+          timestamp:
+            parse_from: attributes.timestamp
+            layout: '%y%m%d %H:%M:%S.%f'
+            location: {{ .timezone }}
+          severity:
+            parse_from: attributes.Type
+    {{ end }} # End of sql schema logs
```
Comment (cpheps) [inline plugins/cockroachdb_logs.yaml:168]: nit: Same here

File: plugins/cockroachdb_logs.yaml [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+      operators:
+        - type: regex_parser
+          regex: '^(?P<Type>[IWEF])(?P<timestamp>\d{6}\s+\d{2}:\d{2}:\d{2}\.\d{6})\s+(?P<GoID>\d+)\s*(?:(?P<Channel>\d+)@)?(?P<Location>\S+)\s+(?:)⋮?\s+(?P<Node>\[[^\]]+\])\s+(?:(?P<Counter>\d+\s+)?)(?P<Message>.*)'
+          timestamp:
+            parse_from: attributes.timestamp
+            layout: '%y%m%d %H:%M:%S.%f'
+            location: {{ .timezone }}
+          severity:
+            parse_from: attributes.Type
+    {{ end }} # End of telemetry logs
```
Comment (cpheps) [inline plugins/cockroachdb_logs.yaml:190]: nit: Same here

File: plugins/cockroachdb_logs.yaml [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+      operators:
+        - type: regex_parser
+          regex: '^(?P<Type>[IWEF])(?P<timestamp>\d{6}\s+\d{2}:\d{2}:\d{2}\.\d{6})\s+(?P<GoID>\d+)\s*(?:(?P<Channel>\d+)@)?(?P<Location>\S+)\s+(?:)⋮?\s+(?P<Node>\[[^\]]+\])\s+(?:(?P<Counter>\d+\s+)?)(?P<Message>.*)'
+          timestamp:
+            parse_from: attributes.timestamp
+            layout: '%y%m%d %H:%M:%S.%f'
+            location: {{ .timezone }}
+          severity:
+            parse_from: attributes.Type
+    {{ end }} # End of kv distribution logs
```
Comment (cpheps) [inline plugins/cockroachdb_logs.yaml:212]: nit: Same here

File: plugins/cockroachdb_logs.yaml [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+      operators:
+        - type: regex_parser
+          regex: '^(?P<Type>[IWEF])(?P<timestamp>\d{6}\s+\d{2}:\d{2}:\d{2}\.\d{6})\s+(?P<GoID>\d+)\s*(?:(?P<Channel>\d+)@)?(?P<Location>\S+)\s+(?:)⋮?\s+(?P<Node>\[[^\]]+\])\s+(?:(?P<Counter>\d+\s+)?)(?P<Message>.*)'
+          timestamp:
+            parse_from: attributes.timestamp
+            layout: '%y%m%d %H:%M:%S.%f'
+            location: {{ .timezone }}
+          severity:
+            parse_from: attributes.Type
+    {{ end }} # End of pebble logs.
```
Comment (cpheps) [inline plugins/cockroachdb_logs.yaml:234]: nit: Same here

File: plugins/cockroachdb_logs.yaml [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+    type: "[]string"
+    default:
+      - "/var/log/cockroach-data/logs/cockroach-pebble.log"
+  - name: start_at
+    description: At startup, where to start reading logs from the file ('beginning' or 'end')
+    type: string
+    supported:
+      - beginning
+      - end
+    default: beginning
```
Comment (cpheps) [inline plugins/cockroachdb_logs.yaml:74]: default should be `end`

File: plugins/cockroachdb_logs.yaml (hunk 3 above)
Comment (cpheps) [reply plugins/cockroachdb_logs.yaml:78]: Can we also add the `parse_to` and `retain_raw_logs` fields like we do [here](https://github.com/observIQ/observiq-otel-collector/blob/release/v1.25.0/plugins/mongodb_logs.yaml#L18-L28)

File: plugins/cockroachdb_logs.yaml [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+      attributes:
+        log_type: 'cockroach.health'
+      operators:
+        - type: regex_parser
+          regex: '^(?P<Type>[IWEF])(?P<timestamp>\d{6}\s+\d{2}:\d{2}:\d{2}\.\d{6})\s+(?P<GoID>\d+)\s*(?:(?P<Channel>\d+)@)?(?P<Location>\S+)\s+(?:)⋮?\s+(?P<Node>\[[^\]]+\])\s+(?:(?P<Counter>\d+\s+)?)(?P<Message>.*)'
+          timestamp:
+            parse_from: attributes.timestamp
+            layout: '%y%m%d %H:%M:%S.%f'
+            location: {{ .timezone }}
+          severity:
```
Comment (cpheps) [inline plugins/cockroachdb_logs.yaml:99]: I think the severities need a mapping. They are showing as unspecified for me. An example is [here](https://github.com/observIQ/observiq-otel-collector/blob/release/v1.25.0/plugins/apache_http_logs.yaml#L74-L81)

**Review by cpheps — CHANGES_REQUESTED** [review]

File: plugins/cockroachdb_logs.yaml [hunk 12]
```diff
@@ -0,0 +1,317 @@
...
+    default: end 
+  - name: timezone
+    description: Timezone to use when parsing the timestamp
+    type: timezone
+    default: UTC
+  - name: retain_raw_logs
+    description: When enabled will preserve the original log message in a `raw_log` key. This will either be in the `body` or `attributes` depending on how `parse_to` is configured.
+    type: bool
+    default: false
+  - name: parse_to
```
Comment (cpheps) [inline plugins/cockroachdb_logs.yaml:83]: Parse too is not actually being used it looks like. See apache combined as an example how it is used to set the initial parsing [here](https://github.com/observIQ/observiq-otel-collector/blob/release/v1.26.0/plugins/apache_combined_logs.yaml#L47) and that changes how you reference fields later on like [here](https://github.com/observIQ/observiq-otel-collector/blob/release/v1.26.0/plugins/apache_combined_logs.yaml#L52)

File: plugins/cockroachdb_logs.yaml [hunk 13]
```diff
@@ -0,0 +1,317 @@
...
+    type: string
+    supported:
+      - beginning
+      - end
+    default: end 
+  - name: timezone
+    description: Timezone to use when parsing the timestamp
+    type: timezone
+    default: UTC
+  - name: retain_raw_logs
```
Comment (cpheps) [inline plugins/cockroachdb_logs.yaml:79]: This is not used either see how it is used here [here](https://github.com/observIQ/observiq-otel-collector/blob/release/v1.26.0/plugins/apache_combined_logs.yaml#L39-L44) to save off the body) then again [here](https://github.com/observIQ/observiq-otel-collector/blob/release/v1.26.0/plugins/apache_combined_logs.yaml#L59-L64) to re add it at the end.

File: plugins/cockroachdb_logs.yaml [hunk 14]
```diff
@@ -0,0 +1,317 @@
...
+    type: bool
+    default: false
+  - name: parse_to
+    description: Where to parse structured log parts
+    type: string
+    supported:
+      - body
+      - attributes
+    default: body
+  - name: offset_storage_dir
```
Comment (cpheps) [inline plugins/cockroachdb_logs.yaml:90]: This needs to be set as an extension before the receivers like [here](https://github.com/observIQ/observiq-otel-collector/blob/release/v1.26.0/plugins/iis_logs.yaml#L72-L74). Then is needs to be referenced in each `filelog` receiver like [here](https://github.com/observIQ/observiq-otel-collector/blob/release/v1.26.0/plugins/iis_logs.yaml#L79)

**Review by cpheps — CHANGES_REQUESTED** [review]

File: plugins/cockroachdb_logs.yaml [hunk 15]
```diff
@@ -0,0 +1,347 @@
...
+      start_at: {{ .start_at }}
+      attributes:
+        log_type: 'cockroach.health'
+      operators:
+        {{ if .retain_raw_logs }}
+        - id: save_raw_log
+          type: copy
+          from: body
+          to: attributes.raw_log
+        {{ end }}
```
Comment (cpheps) [inline plugins/cockroachdb_logs.yaml:117]: This raw logs block and the one at the end to move it if `parse_to` is `body` needs to be on every filelog receiver.

**Review by cpheps — CHANGES_REQUESTED** [review]

**Review by cpheps — APPROVED** [review]

**Pull request merged**